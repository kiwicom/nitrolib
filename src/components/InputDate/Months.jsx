// @flow strict
import * as React from "react";
import getMonth from "date-fns/getMonth";

import Select from "../Select";
import { Consumer } from "../../services/intl/context";

const MONTHS = [
  __("common.months.january"),
  __("common.months.february"),
  __("common.months.march"),
  __("common.months.april"),
  __("common.months.may"),
  __("common.months.june"),
  __("common.months.july"),
  __("common.months.august"),
  __("common.months.september"),
  __("common.months.october"),
  __("common.months.november"),
  __("common.months.december"),
];

type Props = {|
  id: string,
  value: Date,
  onChange: (ev: SyntheticInputEvent<HTMLSelectElement>) => void,
  months: number[],
|};

const Months = ({ id, value, onChange, months }: Props) => (
  <Consumer>
    {({ translate }) => (
      <Select id={`${id}-month`} value={String(getMonth(value))} onChange={onChange}>
        {months.map(month => (
          <option key={month} value={month}>
            {translate(MONTHS[month])}
          </option>
        ))}
      </Select>
    )}
  </Consumer>
);

export default Months;
