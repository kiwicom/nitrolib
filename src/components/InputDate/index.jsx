// @flow strict
import * as React from "react";
import setDate from "date-fns/setDate";
import setMonth from "date-fns/setMonth";
import setYear from "date-fns/setYear";
import Stack from "@kiwicom/orbit-components/lib/Stack";

import calculateRanges from "../../services/dates/calculateRanges";
import Dates from "./components/Dates";
import Months from "./components/Months";
import Years from "./components/Years";

type Props = {|
  value: Date,
  min: Date,
  max: Date,
  format: string[],
  onChange: Date => void,
|};

class InputDate extends React.PureComponent<Props> {
  static defaultProps = {
    format: ["D", "M", "Y"],
  };

  handleChangeDate = (ev: SyntheticInputEvent<HTMLSelectElement>) => {
    const { onChange, value } = this.props;

    onChange(setDate(value, Number(ev.target.value)));
  };

  handleChangeMonth = (ev: SyntheticInputEvent<HTMLSelectElement>) => {
    const { onChange, value } = this.props;

    onChange(setMonth(value, Number(ev.target.value)));
  };

  handleChangeYear = (ev: SyntheticInputEvent<HTMLSelectElement>) => {
    const { onChange, value } = this.props;

    onChange(setYear(value, Number(ev.target.value)));
  };

  render() {
    const { value, min, max, format } = this.props;
    const { dates, months, years } = calculateRanges(min, max, value);

    const parts = {
      D: <Dates value={value} onChange={this.handleChangeDate} dates={dates} />,
      M: <Months value={value} onChange={this.handleChangeMonth} months={months} />,
      Y: <Years value={value} onChange={this.handleChangeYear} years={years} />,
    };

    return (
      <Stack flex spacing="condensed" justify="between">
        {format.map(item => (
          <Stack key={item}>{parts[item]}</Stack>
        ))}
      </Stack>
    );
  }
}

export default InputDate;
