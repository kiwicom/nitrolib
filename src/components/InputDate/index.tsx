import * as React from "react";
import setDate from "date-fns/setDate";
import setMonth from "date-fns/setMonth";
import setYear from "date-fns/setYear";
import Stack from "@kiwicom/orbit-components/lib/Stack";

import calculateRanges from "../../services/dates/calculateRanges";
import Dates from "./components/Dates";
import Months from "./components/Months";
import Years from "./components/Years";

type Props = {
  id: string,
  value: Date,
  min: Date,
  max: Date,
  format: string[],
  onChange: Date => void,
};

class InputDate extends React.PureComponent<Props> {
  static defaultProps = {
    format: ["D", "M", "Y"],
  };

  handleChangeDate = (ev: SyntheticInputEvent<HTMLSelectElement>) => {
    const { onChange, value } = this.props;

    onChange(setDate(value, Number(ev.target.value)));
  };

  handleChangeMonth = (ev: SyntheticInputEvent<HTMLSelectElement>) => {
    const { onChange, value } = this.props;

    onChange(setMonth(value, Number(ev.target.value)));
  };

  handleChangeYear = (ev: SyntheticInputEvent<HTMLSelectElement>) => {
    const { onChange, value } = this.props;

    onChange(setYear(value, Number(ev.target.value)));
  };

  render() {
    const { id, value, min, max, format } = this.props;
    const { dates, months, years } = calculateRanges(min, max, value);

    const parts = {
      D: <Dates id={id} value={value} onChange={this.handleChangeDate} dates={dates} />,
      M: <Months id={id} value={value} onChange={this.handleChangeMonth} months={months} />,
      Y: <Years id={id} value={value} onChange={this.handleChangeYear} years={years} />,
    };

    return (
      <Stack flex spacing="condensed" justify="between">
        {format.map(item => (
          <Stack>{parts[item]}</Stack>
        ))}
      </Stack>
    );
  }
}

export default InputDate;
