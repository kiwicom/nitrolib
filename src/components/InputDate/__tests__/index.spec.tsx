import * as React from "react";
import { shallow, mount } from "enzyme";
import addYears from "date-fns/addYears";
import setDate from "date-fns/setDate";
import setMonth from "date-fns/setMonth";
import setYear from "date-fns/setYear";

import InputDate from "..";

const NOW = new Date(Date.UTC(2018, 2, 1));

describe("#InputDate", () => {
  test("render regular", () => {
    const wrapper = shallow(
      <InputDate
        id="test"
        value={NOW}
        onChange={jest.fn()}
        min={addYears(NOW, -1)}
        max={addYears(NOW, 2)}
      />,
    );

    expect(wrapper.find("Stack").exists()).toBe(true);
  });

  test("render ddmmyyyy", () => {
    const wrapper = mount(
      <InputDate
        id="test"
        value={NOW}
        onChange={jest.fn()}
        min={addYears(NOW, -1)}
        max={addYears(NOW, 2)}
        format={["D", "M", "Y"]}
      />,
    );

    expect(
      wrapper
        .find("Select")
        .at(0)
        .prop("value"),
    ).toEqual("1");

    expect(
      wrapper
        .find("Select")
        .at(1)
        .prop("value"),
    ).toEqual("2");

    expect(
      wrapper
        .find("Select")
        .at(2)
        .prop("value"),
    ).toEqual("2018");
  });

  test("render mmddyyyy", () => {
    const wrapper = mount(
      <InputDate
        id="test"
        value={NOW}
        onChange={jest.fn()}
        min={addYears(NOW, -1)}
        max={addYears(NOW, 2)}
        format={["M", "D", "Y"]}
      />,
    );

    expect(
      wrapper
        .find("Select")
        .at(0)
        .prop("value"),
    ).toEqual("2");

    expect(
      wrapper
        .find("Select")
        .at(1)
        .prop("value"),
    ).toEqual("1");

    expect(
      wrapper
        .find("Select")
        .at(2)
        .prop("value"),
    ).toEqual("2018");
  });

  test("render yyyymmdd", () => {
    const wrapper = mount(
      <InputDate
        id="test"
        value={NOW}
        onChange={jest.fn()}
        min={addYears(NOW, -1)}
        max={addYears(NOW, 2)}
        format={["Y", "M", "D"]}
      />,
    );

    expect(
      wrapper
        .find("Select")
        .at(0)
        .prop("value"),
    ).toEqual("2018");

    expect(
      wrapper
        .find("Select")
        .at(1)
        .prop("value"),
    ).toEqual("2");

    expect(
      wrapper
        .find("Select")
        .at(2)
        .prop("value"),
    ).toEqual("1");
  });

  test("handle change date", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <InputDate
        id="test"
        value={NOW}
        onChange={onChange}
        min={addYears(NOW, -1)}
        max={addYears(NOW, 2)}
      />,
    );

    wrapper.instance().handleChangeDate({ target: { value: "10" } });

    expect(onChange).toBeCalledWith(setDate(NOW, 10));
  });

  test("handle change month", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <InputDate
        id="test"
        value={NOW}
        onChange={onChange}
        min={addYears(NOW, -1)}
        max={addYears(NOW, 2)}
      />,
    );

    wrapper.instance().handleChangeMonth({ target: { value: "10" } });

    expect(onChange).toBeCalledWith(setMonth(NOW, 10));
  });

  test("handle change year", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <InputDate
        id="test"
        value={NOW}
        onChange={onChange}
        min={addYears(NOW, -1)}
        max={addYears(NOW, 2)}
      />,
    );

    wrapper.instance().handleChangeYear({ target: { value: "2020" } });

    expect(onChange).toBeCalledWith(setYear(NOW, 2020));
  });
});
