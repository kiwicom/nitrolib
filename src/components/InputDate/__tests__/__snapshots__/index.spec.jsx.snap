// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#InputDate render ddmmyyyy 1`] = `
<Stack
  flex={true}
  justify="between"
  spacing="condensed"
>
  <Stack>
    <Dates
      dates={
        Array [
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
          11,
          12,
          13,
          14,
          15,
          16,
          17,
          18,
          19,
          20,
          21,
          22,
          23,
          24,
          25,
          26,
          27,
          28,
        ]
      }
      id="test"
      onChange={[Function]}
      value={2018-02-01T00:00:00.000Z}
    />
  </Stack>
  <Stack>
    <Months
      id="test"
      months={
        Array [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
          11,
        ]
      }
      onChange={[Function]}
      value={2018-02-01T00:00:00.000Z}
    />
  </Stack>
  <Stack>
    <Years
      id="test"
      onChange={[Function]}
      value={2018-02-01T00:00:00.000Z}
      years={
        Array [
          2017,
          2018,
          2019,
          2020,
        ]
      }
    />
  </Stack>
</Stack>
`;

exports[`#InputDate render mmddyyyy 1`] = `
<Stack
  flex={true}
  justify="between"
  spacing="condensed"
>
  <Stack>
    <Months
      id="test"
      months={
        Array [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
          11,
        ]
      }
      onChange={[Function]}
      value={2018-02-01T00:00:00.000Z}
    />
  </Stack>
  <Stack>
    <Dates
      dates={
        Array [
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
          11,
          12,
          13,
          14,
          15,
          16,
          17,
          18,
          19,
          20,
          21,
          22,
          23,
          24,
          25,
          26,
          27,
          28,
        ]
      }
      id="test"
      onChange={[Function]}
      value={2018-02-01T00:00:00.000Z}
    />
  </Stack>
  <Stack>
    <Years
      id="test"
      onChange={[Function]}
      value={2018-02-01T00:00:00.000Z}
      years={
        Array [
          2017,
          2018,
          2019,
          2020,
        ]
      }
    />
  </Stack>
</Stack>
`;

exports[`#InputDate render regular 1`] = `
<Stack
  flex={true}
  justify="between"
  spacing="condensed"
>
  <Stack>
    <Dates
      dates={
        Array [
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
          11,
          12,
          13,
          14,
          15,
          16,
          17,
          18,
          19,
          20,
          21,
          22,
          23,
          24,
          25,
          26,
          27,
          28,
        ]
      }
      id="test"
      onChange={[Function]}
      value={2018-02-01T00:00:00.000Z}
    />
  </Stack>
  <Stack>
    <Months
      id="test"
      months={
        Array [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
          11,
        ]
      }
      onChange={[Function]}
      value={2018-02-01T00:00:00.000Z}
    />
  </Stack>
  <Stack>
    <Years
      id="test"
      onChange={[Function]}
      value={2018-02-01T00:00:00.000Z}
      years={
        Array [
          2017,
          2018,
          2019,
          2020,
        ]
      }
    />
  </Stack>
</Stack>
`;

exports[`#InputDate render yyyymmdd 1`] = `
<Stack
  flex={true}
  justify="between"
  spacing="condensed"
>
  <Stack>
    <Years
      id="test"
      onChange={[Function]}
      value={2018-02-01T00:00:00.000Z}
      years={
        Array [
          2017,
          2018,
          2019,
          2020,
        ]
      }
    />
  </Stack>
  <Stack>
    <Months
      id="test"
      months={
        Array [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
          11,
        ]
      }
      onChange={[Function]}
      value={2018-02-01T00:00:00.000Z}
    />
  </Stack>
  <Stack>
    <Dates
      dates={
        Array [
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
          11,
          12,
          13,
          14,
          15,
          16,
          17,
          18,
          19,
          20,
          21,
          22,
          23,
          24,
          25,
          26,
          27,
          28,
        ]
      }
      id="test"
      onChange={[Function]}
      value={2018-02-01T00:00:00.000Z}
    />
  </Stack>
</Stack>
`;
